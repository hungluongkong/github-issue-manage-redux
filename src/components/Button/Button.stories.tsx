import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Button from '.';
import { PreviewArea, Code, DetailTable } from '../commonStyled';
import { ButtonType, ButtonSize } from '../../types';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    {
      name: 'type',
      type: 'enum ButtonType { Default, Primary }',
      default: 'ButtonType.Primary',
    },
    {
      name: 'size',
      type: 'enum ButtonSize { Small, Normal }',
      default: 'ButtonSize.Normal',
    },
    { name: 'value', type: 'string', default: 'ButtonType.Primary' },
    {
      name: 'onClicked',
      type: 'function',
      description:
        '((event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void)',
    },
  ],
});

storiesOf('Button', module)
  .addDecorator((story) => <div style={{ margin: '0 30px' }}>{story()}</div>)
  .add('Button Style', () => (
    <>
      <h3>Button Types</h3>
      <PreviewArea>
        <Button
          type={ButtonType.Primary}
          onClick={action('click')}
          value="Primary Button"
        />
        <Button
          type={ButtonType.Default}
          onClick={action('click')}
          value="Default Button"
        />
      </PreviewArea>
      <Code>
        {`
          <Button type={ButtonType.Primary} onClick={action('click')}>Primary Button</Button>
          <Button type={ButtonType.Default} onClick={action('click')}>Default Button</Button>
        `}
      </Code>
      {propsDetail}
    </>
  ))
  .add('Button Size', () => (
    <>
      <h3>Button Size</h3>
      <PreviewArea>
        <Button
          onClick={action('click')}
          size={ButtonSize.Normal}
          value="Normal Button"
        />
        <Button
          onClick={action('click')}
          size={ButtonSize.Small}
          value="Small Button"
        />
      </PreviewArea>
      <Code>
        {`
          <Button onClick={action('click')} size={ButtonSize.Normal}>Normal Button</Button>
          <Button onClick={action('click')} size={ButtonSize.Small}>Small Button</Button>
        `}
      </Code>
      {propsDetail}
    </>
  ));
