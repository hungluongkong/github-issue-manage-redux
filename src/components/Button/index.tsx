import React from 'react';
import styled from 'styled-components';

import * as color from '../../theme/color';
import * as metrics from '../../theme/metrics';
import { ButtonType, ButtonProps, ButtonSize } from '../../types';

/**
 * Get css style for each type of button.
 * @param type type of button (enum)
 */
const typeToColor = (type: ButtonType): string =>
  type === ButtonType.Primary
    ? `
      background: ${color.PRIMARY_COLOR};
      color: ${color.DEFAULT_COLOR};
    `
    : `
      background: ${color.DEFAULT_COLOR};
      color: ${color.PRIMARY_COLOR};
    `;

/**
 * Styling button by buttonSize
 * @param bSize size of button (enum)
 */
const sizeToSize = (bSize: ButtonSize): string =>
  bSize === ButtonSize.Small
    ? `
      line-height: ${metrics.ButtonSizing.small.lineHeight};
      padding: 3px 5px;
      font-size: ${metrics.FontSize.Small};
    `
    : // Style normal button here
      '';

const StyledButton = styled.button<{ bType: ButtonType; size: ButtonSize }>`
  display: inline-block;
  min-width: 101px;
  text-align: center;
  font-size: ${metrics.FontSize.Default};
  line-height: 20px;
  padding: 6px 10px;
  vertical-align: middle;
  border-radius: 4px;
  border: ${color.PRIMARY_COLOR} 1px solid;

  :hover {
    opacity: ${metrics.OPACITY.hover};
    box-shadow: 0 5px 15px ${color.SHADOW_BORDER};
    cursor: pointer;
  }

  ${(props: { bType: ButtonType }): string => typeToColor(props.bType)}
  ${(props: { size: ButtonSize }): string => sizeToSize(props.size)}
`;

const Button = ({
  type = ButtonType.Primary,
  size = ButtonSize.Normal,
  value = '',
  onClick = () => {},
}: ButtonProps): JSX.Element => (
  <StyledButton
    {...{ size, onClick }}
    // bType: button type
    bType={type}
  >
    {value}
  </StyledButton>
);

export default Button;
