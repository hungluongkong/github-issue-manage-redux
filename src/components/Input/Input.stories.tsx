import React from 'react';
import { storiesOf } from '@storybook/react';
import Input from '.';
import { PreviewArea, Code, DetailTable } from '../commonStyled';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    { name: 'defaultValue', type: 'String' },
    { name: 'placeholder', type: 'String' },
    { name: 'isRequire', type: 'boolean', default: 'false' },
    { name: 'textRef', type: 'React.RefObject<HTMLInputElement>' },
  ],
});

storiesOf('Input', module)
  .add('Default', () => (
    <>
      <h3>Default Input</h3>
      <PreviewArea>
        <Input defaultValue="Text" placeholder="Input here" />
      </PreviewArea>
      <Code>
        {`
          <Input defaultValue="Text" placeholder="Input here" />
        `}
      </Code>
      {propsDetail}
    </>
  ));
