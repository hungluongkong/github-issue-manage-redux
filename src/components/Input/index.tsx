import React from 'react';
import styled from 'styled-components';

import * as color from '../../theme/color';
import { InputProps } from '../../types';

const InputStyled = styled.input`
  padding: 5px 12px;
  background-color: ${color.INPUT_BG_COLOR};
  line-height: 20px;
  border: 1px solid ${color.INPUT_BORDER_COLOR};
  border-radius: 6px;
  box-sizing: border-box;
  flex-grow: 1;
  width: 100%;

  :focus {
    box-shadow: 1px 1px 1px 1px ${color.SHADOW_BORDER};
  }
`;

const Input = ({
  defaultValue = '',
  isRequire = false,
  placeholder = '',
  textRef,
  onChange = () => {},
}: InputProps): JSX.Element => (
  <InputStyled
    placeholder={placeholder}
    defaultValue={defaultValue}
    ref={textRef as React.RefObject<HTMLInputElement>}
    onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
      onChange(e.target.value)
    }
    required={isRequire}
  />
);

export default Input;
