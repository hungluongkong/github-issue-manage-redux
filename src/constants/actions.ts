enum Action {
  GET_ISSUES = 'GET_ISSUES',
  GET_ISSUES_SUCCESS = 'GET_ISSUES_SUCCESS',
  GET_ISSUES_ERROR = 'GET_ISSUES_ERROR',

  UPDATE_ISSUE = 'UPDATE_ISSUE',
  TOGGLE_LOCK_ISSUE = 'TOGGLE_LOCK_ISSUE',
  UPDATE_ISSUE_SUCCESS = 'UPDATE_ISSUE_SUCCESS',

  REPORT_TASK = 'REPORT_TASK',

  TOGGLE_FORM = 'TOGGLE_FORM',
  TOGGLE_ISSUE = 'TOGGLE_ISSUE',
  TOGGLE_ERROR = 'TOGGLE_ERROR',
}

export default Action;
