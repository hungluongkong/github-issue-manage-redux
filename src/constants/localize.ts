
export const en = {
  FORM_TITLE: 'Title',
  FORM_CONTENT_TITLE: 'Content',
  FORM_TITLE_PLACEHOLDER: 'Your title here',
  FORM_CONTENT_PLACEHOLDER: 'Write something about this issue here',

  ISSUE_LIST_TITLE: 'Issue',
  ISSUE_LIST_NO_ISSUE: 'No Issue',

  ISSUE_CREATED_AT: 'Created at ',
  ISSUE_UPDATED_AT: 'Updated at ',

  BTN_ISSUE_LOCK: 'Lock Issue',
  BTN_ISSUE_UNLOCK: 'Unlock Issue',
  BTN_ISSUE_EDIT: 'Edit',
  BTN_ISSUE_ADD: 'New Issue',
  BTN_FORM_CLOSE: 'Close Form',
  BTN_SUBMIT: 'Submit',
  BTN_OK: 'Ok',
  BTN_CONFIRM: 'Confirm',

  WARN_EMPTY_TITLE: 'Title cannot be empty',

  INFORM_LOCK_DONE: 'Issue has locked!',
  INFORM_UNLOCK_DONE: 'Issue has unlocked!',
  INFORM_ADD_DONE: 'Issue added successful!',
  INFORM_UPDATE_DONE: 'Issue has updated!',
  INFORM_ISSUES_GET_DONE: 'Connect successful!',

  ERROR_SOMETHING_WRONG: 'Something went wrong!',
};
