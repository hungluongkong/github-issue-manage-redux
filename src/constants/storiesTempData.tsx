/**
 * Template response json string from github.
 * This is a list of issues.
 * URL: https://api.github.com/repos/hungluong-asnet/demo-api/issues
 */
export const issueJson: string = `
{
  "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1",
  "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
  "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/labels{/name}",
  "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/comments",
  "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/events",
  "html_url": "https://github.com/hungluong-asnet/demo-api/issues/1",
  "id": 653740906,
  "node_id": "MDU6SXNzdWU2NTM3NDA5MDY=",
  "number": 1,
  "title": "isssue no 1",
  "user": {
    "login": "hungluong-asnet",
    "id": 68043715,
    "node_id": "MDQ6VXNlcjY4MDQzNzE1",
    "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/hungluong-asnet",
    "html_url": "https://github.com/hungluong-asnet",
    "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
    "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
    "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
    "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
    "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
    "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
    "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
    "type": "User",
    "site_admin": false
  },
  "labels": [

  ],
  "state": "open",
  "locked": false,
  "assignee": null,
  "assignees": [

  ],
  "milestone": null,
  "comments": 0,
  "created_at": "2020-07-09T03:48:36Z",
  "updated_at": "2020-07-09T03:48:36Z",
  "closed_at": null,
  "author_association": "OWNER",
  "active_lock_reason": null,
  "body": "Nothing inside issue"
}
`;

export const issueListJson: string = `
  [
    {
      "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2",
      "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
      "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/labels{/name}",
      "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/comments",
      "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/events",
      "html_url": "https://github.com/hungluong-asnet/demo-api/issues/2",
      "id": 654622112,
      "node_id": "MDU6SXNzdWU2NTQ2MjIxMTI=",
      "number": 2,
      "title": "#practice #make #api",
      "user": {
        "login": "hungluong-asnet",
        "id": 68043715,
        "node_id": "MDQ6VXNlcjY4MDQzNzE1",
        "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/hungluong-asnet",
        "html_url": "https://github.com/hungluong-asnet",
        "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
        "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
        "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
        "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
        "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
        "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
        "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
        "type": "User",
        "site_admin": false
      },
      "labels": [

      ],
      "state": "open",
      "locked": false,
      "assignee": null,
      "assignees": [

      ],
      "milestone": null,
      "comments": 0,
      "created_at": "2020-07-10T08:47:16Z",
      "updated_at": "2020-07-10T08:47:16Z",
      "closed_at": null,
      "author_association": "OWNER",
      "active_lock_reason": null,
      "body": "**Don't know why this issue is here. and why we need to create it?**"
    },
    ${issueJson}
  ]`;
