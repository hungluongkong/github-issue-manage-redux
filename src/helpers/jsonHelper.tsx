/* eslint-disable camelcase */
import { Issue } from '../types';

export interface RawIssue {
  id: number;
  number: number;
  title: string;
  body: string;
  locked: boolean;
  url: string;
  created_at: string;
  updated_at: string;
}

export const RawToIssue = (raws: RawIssue[]): Issue[] =>
  raws
    .map((raw: RawIssue) => {
      const createdAt = new Date(raw.created_at);
      const updatedAt = new Date(raw.updated_at);

      // Destructor - remove unused properties
      const { id, number, url, title, body, locked } = raw;

      // Create Item
      return {
        id,
        number,
        title,
        body,
        locked,
        url,
        createdAt,
        updatedAt,
      } as Issue;
    })
    .sort((issue: Issue, nextIssue: Issue) => issue.number - nextIssue.number);

/**
 * Convert to list of issue from json
 * @param json response from github api
 */
export const JsonToIssueList = (json: string): Issue[] => {
  const raws: RawIssue[] = JSON.parse(json);

  return RawToIssue(raws);
};
