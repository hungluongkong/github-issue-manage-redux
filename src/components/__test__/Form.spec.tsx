import React from 'react';
import { create } from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import IssueForm from '../Form';
import Button from '../Button';
import Input from '../Input';

import * as localize from '../../constants/localize';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';

describe('Unit test for Issue Form component', () => {
  it('snapshot', () => {
    const buttonComponent = create(<IssueForm onSubmit={() => {}} />);
    expect(buttonComponent).toMatchSnapshot();
  });

  it('should show form', () => {
    const form = shallow(<IssueForm onSubmit={jest.fn()} />);
    expect(form.find(Button).prop('value')).toEqual(localize.en.BTN_SUBMIT);
  });

  it('simulate submit', () => {
    const onSubmit = jest.fn();
    const [issue] = JsonToIssueList(issueListJson);
    const form = shallow(<IssueForm issue={issue} onSubmit={onSubmit} />);

    form.find(Button).simulate('click');
    expect(onSubmit).toBeCalledTimes(1);
  });

  it('should received data on submit', () => {
    const newTitle = 'new value';
    const onSubmit = jest.fn();
    const form = mount(<IssueForm onSubmit={onSubmit} />);
    form.find(Input).getDOMNode<HTMLInputElement>().value = newTitle;
    form.find(Button).simulate('click');
    expect(onSubmit).toBeCalledWith({ body: '', title: newTitle });
  });
});
