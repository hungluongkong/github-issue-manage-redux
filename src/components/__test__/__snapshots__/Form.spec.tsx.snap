// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for Issue Form component snapshot 1`] = `
<div
  className="sc-AxheI kFUkaA"
>
  <div
    className="sc-AxhUy iOlevc"
  >
    <label
      className="sc-AxhCb bnupDl"
    >
      Title
    </label>
    <div
      className="sc-AxgMl cXEjRw"
    >
      <input
        className="sc-AxiKw jxQuPo"
        defaultValue=""
        onChange={[Function]}
        placeholder="Your title here"
        required={false}
      />
    </div>
  </div>
  <div
    className="sc-AxhUy iOlevc"
  >
    <label
      className="sc-AxhCb bnupDl"
    >
      Content
    </label>
    <div
      className="sc-AxgMl fGWvyW"
    >
      <textarea
        className="sc-AxirZ jEyGyt"
        defaultValue=""
        onChange={[Function]}
        required={false}
      />
    </div>
  </div>
  <button
    className="sc-AxjAm caeeVY"
    onClick={[Function]}
    size={1}
  >
    Submit
  </button>
</div>
`;
